src/inc/quic_platform_winkernel.h

```diff
@@ -314,17 +314,14 @@ typedef EX_PUSH_LOCK CXPLAT_RW_LOCK;
 #define CxPlatRwLockReleaseShared(Lock) ExReleasePushLockShared(Lock); KeLeaveCriticalRegion()
 #define CxPlatRwLockReleaseExclusive(Lock) ExReleasePushLockExclusive(Lock); KeLeaveCriticalRegion()
 
-typedef struct CXPLAT_DISPATCH_RW_LOCK {
-    EX_SPIN_LOCK SpinLock;
-    KIRQL PrevIrql;
-} CXPLAT_DISPATCH_RW_LOCK;
+typedef EX_SPIN_LOCK CXPLAT_DISPATCH_RW_LOCK;
 
-#define CxPlatDispatchRwLockInitialize(Lock) (Lock)->SpinLock = 0
+#define CxPlatDispatchRwLockInitialize(Lock) *(Lock) = 0
 #define CxPlatDispatchRwLockUninitialize(Lock)
-#define CxPlatDispatchRwLockAcquireShared(Lock) (Lock)->PrevIrql = ExAcquireSpinLockShared(&(Lock)->SpinLock)
-#define CxPlatDispatchRwLockAcquireExclusive(Lock) (Lock)->PrevIrql = ExAcquireSpinLockExclusive(&(Lock)->SpinLock)
-#define CxPlatDispatchRwLockReleaseShared(Lock) ExReleaseSpinLockShared(&(Lock)->SpinLock, (Lock)->PrevIrql)
-#define CxPlatDispatchRwLockReleaseExclusive(Lock) ExReleaseSpinLockExclusive(&(Lock)->SpinLock, (Lock)->PrevIrql)
+#define CxPlatDispatchRwLockAcquireShared(Lock) KIRQL RwLockPrevIrql = ExAcquireSpinLockShared(Lock)
```

Comment:
This requires CxPlatDispatchRwLockAcquireShared to be called in the same scope as CxPlatDispatchRwLockReleaseShared? Is that good?

Reply:
Can we change the code to not acquire the shared locked if the caller already holds the exclusive lock? Like adding a param to QuicLookupFindConnectionByLocalCidInternal?

Reply:
I feel like these macros with hidden variables are intrinsically dangerous. Why not just make up a cross-platform lock handle?

Reply:
E.g. `CxPlatDispatchRwLockAcquireShared(CXPLAT_LOCK *Lock, CXPLAT_LOCK_HANDLE *LockHandle);`

Reply:
I am trying to not add extra stuff that isn't needed on any other platform.

Reply:
IMO, let the compiler or preprocessor optimize away unused variables, rather than hide them in macros. As a programmer, I would not have expected CxPlatDispatchRwLockAcquireShared to declare a hidden variable.